src/build.cc

```diff
@@ -692,7 +692,7 @@ bool Builder::StartEdge(Edge* edge, string* err) {
   Node* most_recent_input = NULL;
   for (vector<Node*>::iterator i = edge->inputs_.begin();
        i != edge->inputs_.end() - edge->order_only_deps_; ++i) {
-    if (!(*i)->Stat(disk_interface_, err))
+    if (!(*i)->StatIfNecessary(disk_interface_, err))
```

Comment:
This was discussed in #1753 but I think the performance tradeoff should lean toward speed over correctness. If we err on the side of speed, we just might have unnecessary rebuilds on future runs (as opposed to the alternative which is thinking outputs are up to date when they are actually dirty).

Reply:
A build system should not re-run commands when not needed.  We don't know how many downstream dependencies might need to rebuild unnecessarily.

Reply:
We discussed this on the original PR, I think you actually wondered why this wasn't `StatIfNecessary()` to begin with.

This shouldn't cause ninja to rerun commands if the inputs don't change, which is identical behavior to what was here before #1753.

Reply:
Here is an example that change affects:

```bash
rm -f .ninja_deps .ninja_log out1 out2 in1 in2imp

>in1
echo >build.ninja '
rule touch_out_discovered_during_build
  command = touch $out $discovered

rule touch_dep_discovered_during_build
  command = touch $out; echo "$out: $discovered" > $depfile

build out1: touch_out_discovered_during_build in1
  discovered = in2imp

build out2: touch_dep_discovered_during_build in1 || out1
  discovered = in2imp
  depfile = out2.d
  deps = gcc

default out2'

ninja >/dev/null 2>&1

sleep 1
>in1
ninja
ninja -d explain
```

With `Stat`, or prior to #1753, the output is:

```
[1/2] touch out1 in2imp
[2/2] touch out2; echo "out2: in2imp" > out2.d
ninja: no work to do.
```

With `StatIfNecessary` after #1753, the output is:

```
[1/2] touch out1 in2imp
[2/2] touch out2; echo "out2: in2imp" > out2.d
ninja explain: recorded mtime of out2 older than most recent input in2imp (... vs ...)
[1/1] touch out2; echo "out2: in2imp" > out2.d
```

Reply:
Yeah, this test is unfortunate. Using `StatIfNecessary()` could lead to infinite rebuild loops (target would never be clean) with this specific relationship of rules and implicit dependencies being updated by the rules. `out1` has no specific knowledge that it updated `in2imp` so there's nothing we can do when finishing the command for `out1` to update the cached mtime for `in2imp`.

If we want to keep the better build correctness that #1753 provides, it looks like there's no way to avoid `Stat()` when edges are started which does have a performance penalty.

Reply:
IIUC #1753 provides robustness against edits made to the source tree *during the build* by processes *outside the build system*.  We need to decide if supporting that is worth making everyone pay for the extra `Stat` on every build just in case someone does that.

Reply:
This investigation led me to find that [CMake Issue 20025](https://gitlab.kitware.com/cmake/cmake/-/issues/20025#note_925598) has the same root cause, even with Ninja 1.10 (which does not have #1753).  Cases like this might be resolvable using a `dyndep` binding to declare the dynamically-discovered implicit output so that Ninja knows about it.  I still need to try that out to be sure.

However, CMake's cases are just examples of existing cases that may exist in the wild.  Many projects would have had no choice but to express things as in my above example because `dyndep` support is recent.

Reply:
I suggest dropping commit `53f96fed` for now so we can get everything else fixed.  Then open an issue to discuss this optimization.

Reply:
Yeah, that sounds good. I'm going to go ahead and add regression tests for this case anyhow.

Reply:
> might be resolvable using a dyndep binding... I still need to try that out to be sure.

Unfortunately that won't work without additional changes to Ninja.  See the [explanation](https://gitlab.kitware.com/cmake/cmake/-/issues/20025#note_926850) I posted in the CMake issue.